tensorflow/python/kernel_tests/conv_ops_3d_test.py

```diff
@@ -81,22 +94,37 @@ def _VerifyValues(self, tensor_in_sizes, filter_in_sizes, stride, padding,
                     expected):
     results = []
     for data_format, use_gpu in GetTestConfigs():
-      result = self._SetupValuesForDevice(
-          tensor_in_sizes,
-          filter_in_sizes,
-          stride,
-          padding,
-          data_format,
-          use_gpu=use_gpu)
-      results.append(result)
-      tolerance = 1e-2 if use_gpu else 1e-5
+      for dtype in self._DtypesToTest(use_gpu):
+        result = self._SetupValuesForDevice(
+            tensor_in_sizes,
+            filter_in_sizes,
+            stride,
+            padding,
+            data_format,
+            dtype,
+            use_gpu=use_gpu)
+        results.append((result, use_gpu, dtype, data_format))
+
       with self.test_session() as sess:
-        values = sess.run(results)
-        for value in values:
+        tensor_results = [result[0] for result in results]
+        test_params = [result[1:] for result in results]
+        values = sess.run(tensor_results)
+        for i in range(len(tensor_results)):
+          value = values[i]
           print("expected = ", expected)
           print("actual = ", value)
-          self.assertAllClose(expected, value.flatten(), atol=tolerance,
-                              rtol=1e-6)
+          tol = 1e-6
+          if value.dtype == np.float16:
+            tol = 1e-3
+            # fp16 using may result in inf values and large absolute errors
+            # when used with large numbers (fp16 max is 65004.)
+            if np.any(value > 1e4):
```

Comment:
As I stated before, this will skip some important tests, causing some things not to be tested. testConv3D2x2x2FilterStride2, testConv3DStride3, and testConv3D2x2x2FilterStride2Same and not run at all with this, so you should add more tests.

@mjanusz, how were the expected outputs calculated for the original tests? It is a pain to add new tests since new expected outputs need to be calculated.

Reply:
I made normalization of input and supposed  that current fp64 convs implemented correctly => I took them as reference value. It's not a good idea to test fp64 outputs by themselves but I not sure how to solve this problem in different way. 
It's better just to have reference implementation and just compare outputs but it's additional code and we need more tests for tests, yes? :)